contracts/integrations/utils/AdapterBase.sol

```diff
@@ -0,0 +1,178 @@
+        }
+
+        // Execute call
+        _;
+
+        // Transfer incoming assets back to fund
+        for (uint256 i = 0; i < incomingAssets.length; i++) {
+            uint256 postCallAmount = IERC20(incomingAssets[i]).balanceOf(address(this));
+            // Allow incoming asset balance diff to be 0 or negative in case adapters
+            // can send assets directly to the Vault
+            if (postCallAmount <= incomingAssetPreCallAmounts[i]) continue;
+
+            __safeTransfer(
+                incomingAssets[i],
+                msg.sender,
+                sub(postCallAmount, incomingAssetPreCallAmounts[i])
+            );
+        }
+
+        // If excess spend assets, send back to fund
+        for (uint256 i = 0; i < spendAssets.length; i++) {
+            uint256 postCallAmount = IERC20(spendAssets[i]).balanceOf(address(this));
+            if (postCallAmount > spentAssetPreCallAmounts[i]) {
+                __safeTransfer(
+                    spendAssets[i],
+                    msg.sender,
+                    sub(postCallAmount, spentAssetPreCallAmounts[i])
+                );
+            }
+        }
+    }
+
+    modifier onlyVault {
+        require(__isVault(msg.sender), "Only Vault can make this call");
+        _;
+    }
+
+    constructor(address _registry) public {
+        REGISTRY = _registry;
+    }
+
+    // INTERNAL FUNCTIONS
+
+    /// @dev Aggregates a list of _assets and _amounts, removing duplicate assets and summing their amounts.
+    /// At present, this is only used to aggregate fees in integration adapters.
+    /// e.g., in 0x v3, if the takerFee asset is WETH, then takerFee and protocolFee are aggregated.
+    /// It remove assets with an empty address or a balance of 0, so adapters can be dumb.
+    function __aggregateAssets(address[] memory _assets, uint256[] memory _amounts)
+        internal
+        pure
+        returns (
+            address[] memory aggregatedAssets_,
+            uint256[] memory aggregatedAssetAmounts_
+        )
+    {
+        // Get count of unique assets with non-zero values
+        uint256 aggregatedAssetsCount;
+        for (uint256 i = 0; i < _assets.length; i++) {
+            // Ignore assets with a 0 address, 0 amount
+            if (_assets[i] == address(0) || _amounts[i] == 0) continue;
+
+            // Ignore assets that have already been added
+            bool assetAdded;
+            for (uint256 j = 0; j < i; j++) {
+                // Prev asset in array must have non-zero amount
+                if (_assets[i] == _assets[j] && _amounts[j] > 0) {
+                    assetAdded = true;
+                    break;
+                }
+            }
+            if (!assetAdded) aggregatedAssetsCount++;
+        }
+        aggregatedAssets_ = new address[](aggregatedAssetsCount);
+        aggregatedAssetAmounts_ = new uint256[](aggregatedAssetsCount);
+        if (aggregatedAssetsCount == 0) return (aggregatedAssets_, aggregatedAssetAmounts_);
```

Comment:
How do you feel about single-line if statements in general? I usually prefer to never write body-less control statements. E.g. instead of this I'd always prefer: 

```solidity
if (aggregatedAssetsCount == 0) {
  return (aggregatedAssets_, aggregatedAssetAmounts_);
}
```

Simply because an `if` in the code immediately makes my eyes jump to the next line.

Reply:
I'll create a card for updating the style guide now that there are a couple items in this PR, and we can discuss that there and then refactor.